editors/code/src/inlay_hints.ts

```diff
@@ -163,16 +164,21 @@ class HintsUpdater {
 
         this.pending.set(documentUri, tokenSource);
         try {
-            return await sendRequestWithRetry<InlayHint[] | null>(
+            return await sendRequestWithRetry<InlayHint[]>(
                 client,
                 'rust-analyzer/inlayHints',
                 request,
                 tokenSource.token,
             );
-        } finally {
-            if (!tokenSource.token.isCancellationRequested) {
-                this.pending.delete(documentUri);
+        } catch (err) {
+            if (!isCancellationError(err)) {
+                // FIXME: log the error
+                throw err;
             }
+
+            assert(tokenSource.token.isCancellationRequested);
+            this.pending.delete(documentUri);
```

Comment:
Wait, I think this is a bug, we shouldn't delete the documentUri here, shouldn't we?
I.e. the stack frame that canceled this request already override the documentUri entry with new request in this point of time here, right?
@matklad

Reply:
I guess something like the following should be the correct code:
```ts
if (this.pending.get(documentUri == tokenSource)) {
    this.pending.delete(documentUri);
}
```

Reply:
Let me guess, this is why we have time outs. This is because the requests that need to be canceled are actually removed from `this.pending.delete` map!!!

Reply:
Yes, I checked and we do remove the cancellation token source that should stay there...
![image](https://user-images.githubusercontent.com/36276403/75097919-d7239c00-55b8-11ea-88fe-2dda615e39aa.png)